core/client/src/main/java/alluxio/client/file/UnderFileSystemFileInStream.java

```diff
@@ -186,20 +189,27 @@ private void checkIfClosed() {
    */
   private int directRead(byte[] b, int off, int len) throws IOException {
     int bytesLeft = len;
+    int bytesRead = 0;
+    int offset = off;
     while (bytesLeft > 0) {
-      ByteBuffer data = mReader.read(mAddress, mUfsFileId, mPos, bytesLeft);
+      // mPos shouldn't be modified, so keep track of our updated position
+      long currentPosition = mPos + bytesRead;
+
+      ByteBuffer data = mReader.read(mAddress, mUfsFileId, currentPosition, bytesLeft);
```

Comment:
I don't think we can avoid it since I think we would have to do the copy in the NettyReader then. I think if we use bytebuffers throughout we can avoid copying the data.